File: 76. Minimum Window Substring/solution.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        # edge case
        if not s or not t:
            return ""
        
        sol = ""
        seen = {}
        
        # get character count
        chars = set(t)
        req_count = [0]*128
        for u in t:
            i = ord(u)
            req_count[i] += 1
        
        i = 0
        j = 0
        act_count = [0]*128
        while j < len(s):
            # move j
            ind = ord(s[j])
            act_count[ind] += 1 
            
            # move i
            ind = ord(s[i])
            act_count[ind] -= 1
            i += 1
            while self.isValidWindow(req_count, act_count):
                ind = ord(s[i])
                act_count[ind] -= 1
                i += 1
                
            i -= 1
            ind = ord(s[i])
            act_count[ind] += 1
            
            # check if this is a valid window and update sol
            if self.isValidWindow(req_count, act_count):
                if not sol:
                    sol = s[i:j+1]
                else:
                    if len(sol) > j - i + 1:
                        sol = s[i:j+1]
            
            j += 1
        return sol
    
    def isValidWindow(self, required, actual):
        for i in range(128):
            if required[i] > actual[i]:
                return False
        return True

File: 76. Minimum Window Substring/solution.java
class Solution {
    public String minWindow(String s, String t) {
        // edge cases
        if (t.length() == 0 || s.length() == 0 || t.length() > s.length()) {
            return new String("");
        }
        
        if (t.equals(s)) {
            return t;
        }
        
        // init
        Count count = new Count(t);
        int i = 0;
        int j = 0;
        
        // get at least first window
        while (!count.isSafeState() && j < s.length()) {
            count.add(s.charAt(j));
            j++;
        }
        
        if (!count.isSafeState() && j == s.length()) {
            return new String("");
        }
        
        // move window tail
        while (count.isSafeToRemove(s.charAt(i))) {
            count.remove(s.charAt(i));
            i++;
        }
        
        int minI = i;
        int minJ = j;
        
        // keep the window moving
        while (j < s.length()) {
            // move window head
            count.add(s.charAt(j));
            j++;
                            
            // move window tail
            while (count.isSafeToRemove(s.charAt(i))) {
                count.remove(s.charAt(i));
                i++;
            }

            // check if the window is minimum
            if (j - i < minJ - minI) {
                minJ = j;
                minI = i;
            }
        }
        
        return s.substring(minI, minJ);
    }
    
    class Count {
        Map<Character, Integer> currentCount;
        Map<Character, Integer> reqCount;
        
        public Count(String t) {
            currentCount = new HashMap<Character, Integer>();
            reqCount = new HashMap<Character, Integer>();

            char c;
            int count;
            for(int i = 0; i < t.length(); i++) {
                c = t.charAt(i);
                if (reqCount.containsKey(c)) {
                    count = reqCount.get(c);
                    reqCount.put(c, count+1);
                } else {
                    reqCount.put(c, 1);
                    currentCount.put(c, 0);
                }
            }
        }
        
        private boolean isCharImp(char c) {
            return reqCount.containsKey(c);
        }
        
        public void add(char c) {
            if (this.isCharImp(c)) {
                int count = currentCount.get(c);
                currentCount.put(c, count+1);
            }
        }
        
        public void remove(char c) {
            if (this.isCharImp(c)) {
                int count = currentCount.get(c);
                currentCount.put(c, count-1);
            }
        }
        
        public boolean isSafeState() {
            for (char c : reqCount.keySet()) {
                if (reqCount.get(c) > currentCount.get(c)) {
                    return false;
                }
            }
            return true;
        }
        
        public boolean isSafeToRemove(char c) {
            if (!this.isCharImp(c)) {
                return true;
            }
            
            if (currentCount.get(c) - 1 < reqCount.get(c)) {
                return false;
            }
            
            return true;
        }
    }
}
